File: src/ros2_bridge/mappers/trajectory_msgs/joint_trajectory.rs
//! Typed mapper for `trajectory_msgs/msg/JointTrajectory`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn joint_trajectory_to_bus(
    msg: ros_env::trajectory_msgs::msg::JointTrajectory,
) -> crate::trajectory_msgs::msg::v1::JointTrajectory {
    crate::trajectory_msgs::msg::v1::JointTrajectory {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        joint_names: crate::ros2_bridge::mappers::convert::string_seq(msg.joint_names),
        points: msg.points.into_iter().map(crate::ros2_bridge::mappers::trajectory_msgs::joint_trajectory_point::joint_trajectory_point_to_bus).collect(),
    }
}

pub(crate) fn joint_trajectory_to_ros(
    bus: crate::trajectory_msgs::msg::v1::JointTrajectory,
) -> ros_env::trajectory_msgs::msg::JointTrajectory {
    ros_env::trajectory_msgs::msg::JointTrajectory {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        joint_names: crate::ros2_bridge::mappers::convert::ros_string_seq(bus.joint_names),
        points: bus.points.into_iter().map(crate::ros2_bridge::mappers::trajectory_msgs::joint_trajectory_point::joint_trajectory_point_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct TrajectoryMsgsJointTrajectoryMapper;

impl TypedTopicMapper for TrajectoryMsgsJointTrajectoryMapper {
    type Ros = ros_env::trajectory_msgs::msg::JointTrajectory;
    type Bus = crate::trajectory_msgs::msg::v1::JointTrajectory;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(joint_trajectory_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(joint_trajectory_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/trajectory_msgs/joint_trajectory.py
"""Generated mapper for `trajectory_msgs/msg/JointTrajectory`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.trajectory_msgs.joint_trajectory_point import joint_trajectory_point_to_bus, joint_trajectory_point_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.trajectory_msgs.msg.v1 import JointTrajectory as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def joint_trajectory_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.joint_names.extend([str(x) for x in msg.joint_names])
    bus.points.extend([joint_trajectory_point_to_bus(x) for x in msg.points])
    return bus


def joint_trajectory_to_ros(bus):
    from trajectory_msgs.msg import JointTrajectory as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.joint_names = [str(x) for x in bus.joint_names]
    out.points = [joint_trajectory_point_to_ros(x) for x in bus.points]
    return out


class TrajectoryMsgsJointTrajectoryMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from trajectory_msgs.msg import JointTrajectory as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return joint_trajectory_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return joint_trajectory_to_ros(bus)
